java/dataset/src/main/cpp/jni_wrapper.cc

```diff
@@ -470,12 +471,37 @@ JNIEXPORT jlong JNICALL Java_org_apache_arrow_dataset_jni_JniWrapper_createScann
   std::shared_ptr<arrow::dataset::ScannerBuilder> scanner_builder =
       JniGetOrThrow(dataset->NewScan());
   JniAssertOkOrThrow(scanner_builder->Pool(pool));
-  if (columns != nullptr) {
-    std::vector<std::string> column_vector = ToStringVector(env, columns);
+  if (columns_subset != nullptr) {
+    std::vector<std::string> column_vector = ToStringVector(env, columns_subset);
     JniAssertOkOrThrow(scanner_builder->Project(column_vector));
   }
+  if (columns_to_produce_or_filter != nullptr) {
+    auto *buff = reinterpret_cast<jbyte*>(env->GetDirectBufferAddress(columns_to_produce_or_filter));
+    int length = env->GetDirectBufferCapacity(columns_to_produce_or_filter);
+    std::shared_ptr<arrow::Buffer> buffer = JniGetOrThrow(arrow::AllocateBuffer(length));
+    std::memcpy(buffer->mutable_data(), buff, length);
+    arrow::engine::BoundExpressions bounded_expression =
+      JniGetOrThrow(arrow::engine::DeserializeExpressions(*buffer));
+    std::vector<arrow::compute::Expression> project_exprs;
+    std::vector<std::string> project_names;
+    arrow::compute::Expression filter_expr;
+    int filter_count = 0;
+    for(arrow::engine::NamedExpression named_expression : bounded_expression.named_expressions) {
+      if (named_expression.expression.type()->id() == arrow::Type::BOOL) {
+        if (filter_count > 0) {
+          JniThrow("The process only support one filter expression declared");
+        }
+        filter_expr = named_expression.expression;
+        filter_count++;
```

Comment:
You can track this instead with `optional<Expression> filter_expr`

Reply:
That said, we should design the API to separate project and filter expressions in the first place.

Reply:
added